File: app/src/main/java/br/com/example/hamburgos/request/Presenter.java
package br.com.example.hamburgos.request;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import br.com.example.hamburgos.model.Snack;
import br.com.example.hamburgos.ui.MainActivity;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class Presenter {

    private Repository repository;
    private MainActivity activity;

    public Presenter(MainActivity activity) {
        this.activity = activity;
        this.repository = new Repository(activity);
    }


    public void getSnacks() {
        repository.listSnacks(new Callback<List<Snack>>() {
            @Override
            public void onResponse(@NonNull Call<List<Snack>> call, @NonNull Response<List<Snack>> response) {
                if (activity != null) {
                    List<Snack> snacks = response.body();
                    activity.setContent(snacks);
                }
            }

            @Override
            public void onFailure(@NonNull Call<List<Snack>> call, Throwable t) {
                activity.setError("Error occur!");
            }
        });
    }
}

File: app/src/main/java/br/com/example/hamburgos/request/Presenter.kt
package br.com.example.hamburgos.request

import br.com.example.hamburgos.model.Ingredient
import br.com.example.hamburgos.model.Order
import br.com.example.hamburgos.model.Snack
import br.com.example.hamburgos.ui.MainActivity
import br.com.example.hamburgos.util.Constants
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class Presenter(private val activity: MainActivity?) {

    private val repository: Repository = Repository(activity)

    fun getSnacks() {
        repository.listSnacks(object : Callback<List<Snack>> {
            override fun onResponse(call: Call<List<Snack>>, response: Response<List<Snack>>) {
                val snacks = response.body()
                if (snacks != null) {
                    getIngredientsBySnack(snacks)
                }
            }

            override fun onFailure(call: Call<List<Snack>>, t: Throwable) {
                activity?.setError(t.message)
            }
        })
    }

    private fun getIngredientsBySnack(snacks: List<Snack>) {
        for (snack in snacks) {
            repository.getIngredientBySnack(snack.id, object : Callback<List<Ingredient>> {
                override fun onResponse(call: Call<List<Ingredient>>, response: Response<List<Ingredient>>) {
                    val ingredients = response.body()
                    snack.ingredientList = ingredients
                    activity?.setContent(snacks)
                }

                override fun onFailure(call: Call<List<Ingredient>>, t: Throwable) {}
            })
        }
    }

    fun addRequest(snack: Snack) {
        repository.addOrder(snack, object : Callback<Order> {
            override fun onResponse(call: Call<Order>, response: Response<Order>) {
                val request = response.body()

                if (activity != null) {
                    if (request != null) {
                        activity.requestConfirmation()
                    } else {
                        activity.setError(Constants.ORDER_NOT_COMPLETED)
                    }
                }
            }

            override fun onFailure(call: Call<Order>, t: Throwable) {
                activity?.setError(t.message)
            }
        })
    }
}
